File: clients/web/src/lib/api/grant.ts
import { getGrantService } from "@/lib/wasm-core";

export interface ResourceGrant {
  id: number;
  resource_type: string;
  resource_id: string;
  user_id: number;
  granted_by: number;
  created_at: string;
  user?: { id: number; email: string; username: string; name?: string };
  granted_by_user?: { id: number; email: string; username: string; name?: string };
}

export const grantApi = {
  list: async (resourceType: string, resourceId: string) => {
    const json = await getGrantService().list(resourceType, resourceId);
    return JSON.parse(json) as { grants: ResourceGrant[] };
  },

  grant: async (resourceType: string, resourceId: string, userId: number) => {
    const json = await getGrantService().grant(resourceType, resourceId, BigInt(userId));
    return JSON.parse(json) as { grant: ResourceGrant };
  },

  revoke: async (resourceType: string, resourceId: string, grantId: number) => {
    await getGrantService().revoke(resourceType, resourceId, BigInt(grantId));
    return { message: "ok" };
  },
};

File: clients/core/crates/types/src/grant.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceGrantUserBrief {
    pub id: i64,
    pub email: String,
    pub username: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceGrant {
    pub id: i64,
    pub resource_type: String,
    pub resource_id: String,
    pub user_id: i64,
    pub granted_by: i64,
    pub created_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<ResourceGrantUserBrief>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub granted_by_user: Option<ResourceGrantUserBrief>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceGrantListResponse {
    pub grants: Vec<ResourceGrant>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateResourceGrantRequest {
    pub user_id: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceGrantResponse {
    pub grant: ResourceGrant,
}
